### ISSUE TITLE
Embargo asset support?

### ISSUE THREAD

**Issue by nholik** [contentful/contentful.net#267]
I was trying to figure out if there was support for some of the functions related to [embargoed assets](https://www.contentful.com/help/embargoed-assets-overview/).  However, it looks like, there is not support that I can see?  Referring to [this document](https://www.contentful.com/developers/docs/tutorials/general/embargoed-assets-getting-started/) I was hoping to leverage the library for methods equivalent to the asset token generation and asset url signing on that page.

I'd be happy to take a stab at implementing these if you're open to expanding the library to supporting this, unless i'm missing it.  Below are the JS snippets from that page for reference.

I was thinking two methods on the client (roughly):
```csharp
Task<EmbargoAssetKey> CreateEmbargoAssetKey() //would use token on Client (no caching) 
Task<string> SignAssetUrl(string url, string secret, string policy, DateTime expires) 
//EmbargoAssetKey models response
```

```javascript
// sign-url.js, part 2/5

const assetKeyCache = new Map()

/**
 * Creates an asset key with a cache. If an asset key is already available, it will
 * be returned immediately. Otherwise, a fresh asset key will be created and cached.
 * Asset keys are always fetched with the greated possible expiry. This allows
 * multiple requests for shorter expiries to be collapsed into one request.
 *
 * @param host string the Contentful API host to contact
 * @param accessToken string the Contentful API access token to use
 * @param spaceId string the space ID for the asset key
 * @param environmentId string the environment ID for the asset key
 * @param minExpiresAtMs number a JS Unix timestamp (in ms) that is the minimum expiry
 *     time that you require for signing
 * @returns a Promise resolving to an asset key with `policy` and `secret` properties
 */
function createCachedAssetKey(host, accessToken, spaceId, environmentId, minExpiresAtMs) {
    const cacheKey = `${host}:${spaceId}:${environmentId}`
    let cacheItem = assetKeyCache.get(cacheKey)
    if (!cacheItem || cacheItem.expiresAtMs < minExpiresAtMs) {
        // Create a new key at the maximum validity, 48h
        const expiresAtMs = Date.now() + 48 * 60 * 60 * 1000
        if (minExpiresAtMs > expiresAtMs) {
          throw new Error(`Cannot fetch an asset key so far in the future: ${minExpiresAtMs} > ${expiresAtMs}`)
        }
        const promise = createAssetKey(
            host,
            accessToken,
            spaceId,
            environmentId,
            expiresAtMs
        ).catch(err => {
            // If we encounter an error, make sure to clear the cache item if
            // this is the most recent fetch.
            const curCacheItem = assetKeyCache.get(cacheKey)
            if (curCacheItem === cacheItem) {
                assetKeyCache.delete(cacheKey)
            }
            return Promise.reject(err)
        })
        cacheItem = { expiresAtMs, promise }
        assetKeyCache.set(cacheKey, cacheItem)
    }
    return cacheItem.promise
}

const createCachedDeliveryAssetKey = (...args) => createCachedAssetKey('cdn.contentful.com', ...args)
const createCachedManagementAssetKey = (...args) => createCachedAssetKey('api.contentful.com', ...args)
const createCachedPreviewAssetKey = (...args) => createCachedAssetKey('preview.contentful.com', ...args)
```

```javascript
// sign-url.js, part 3/5

// This example uses auth0's jsonwebtoken library. You're free
// to use any JWT library you'd like.
const jwt = require('jsonwebtoken')

/**
 * Generates a signed "token" for an embargoed asset
 *
 * @param secret string the secret retrieved from the asset_keys endpoint
 * @param urlWithoutQueryParams string a url, without query parameters
 *   attached, that you'd like to sign
 * @param expiresAtMs number (optional) the JS Unix timestamp (in ms) of
 *   when the signed URL should expire
 * @returns string the signed url
 */
function generateSignedToken(secret, urlWithoutQueryParams, expiresAtMs) {
    // Convert expiresAtMs to seconds, if defined
    const exp = expiresAtMs ? Math.floor(expiresAtMs / 1000) : undefined
    return jwt.sign({
        sub: urlWithoutQueryParams,
        exp,
    }, secret, { algorithm: 'HS256' })
}
```

**Comment by Roblinde** [comment]
Hi @nholik No, there's no support for that yet. If you wish to have a go at the implementation that'd be awesome. I'll be available if you have any questions. Otherwise I'll probably get around to adding support in the next couple of weeks.

**Comment by nholik** [comment]
Thanks @Roblinde I'll probably look this weekend and give you a PR to look at.

**Comment by nholik** [comment]
@Roblinde  I've put in a PR for this, but some of the functionality, probably deserves more discussion.  What I added is part of the delivery/preview API, so it fits naturally. 

 However, some of the functions related to signing seem like either they would beyond the scope of this API, or I would need some further input on how to add it in.  It doesn't seem like there are general helper methods of this sort - though I would find it quite useful.  Further, doing the url signing would mean adding a JWT library, which the standard microsoft one could be added, but I don't want to take too much liberty.

Again, I'd be happy to submit an additional PR for some of the convenience functions after getting some feedback/guidance on how to fit some of this stuff in.  It doesn't really apply to the major classes for interacting with the delivery+management APIs.

**Comment by Roblinde** [comment]
@nholik thanks for the PR.

I agree that the signing part of the workflow might be something outside of the scope of the SDK. We could potentially expose some helper methods, but I feel like URL-signing and transformation of the asset URL:s would be something that should probably be a separate library or even just a guide for an example .net back end that could be extended and modified to suit the needs of the application (similar to the example express guide here: https://www.contentful.com/developers/docs/tutorials/general/embargoed-assets-getting-started/).

Happy to hear your thoughts on this and see what functionality would be appropriate and helpful in the SDK.

**Comment by nholik** [comment]
@Roblinde yes, after spending more time with the library, only the key generation makes sense and is a natural fit.  I completely agree that the additional helper functions, at least as far as related to this issue, make sense as a separate library, which I may take on for this particular situation.

**Comment by Roblinde** [comment]
@nholik thank you for the contribution. Your PR is now merged to version 7.0.0
